cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+	groupKeys := decisionGroupsToGroupKeys(strategy.Progressive.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
+
+	rolloutresult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(rolloutresult.ClustersToRollout) > 0 {
+		return &strategy, rolloutresult, nil
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.Progressive.Timeout.Timeout)
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// calculate length
+	totalClusters, _, _ := r.pdTracker.ExistingBesides([]clusterv1beta1.GroupKey{})
+	length, err := calculateLength(strategy.Progressive.MaxConcurrency, len(totalClusters))
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// upgrade the rest clusters
+	restClusters, _, restClusterGroups := r.pdTracker.ExistingBesides(clusterGroupKeys)
+	restClustersToGroupKey := r.pdTracker.ClusterToGroupKey(restClusterGroups)
+	rolloutresult = progressivePerCluster(restClusters.UnsortedList(), restClustersToGroupKey, length, failureTimeout, statusFunc)
+
+	return &strategy, rolloutresult, nil
+}
+
+func (r *RolloutHandler) getProgressivePerGroupClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	strategy := RolloutStrategy{Type: ProgressivePerGroup}
+	strategy.ProgressivePerGroup = rolloutStrategy.ProgressivePerGroup.DeepCopy()
+	if strategy.ProgressivePerGroup == nil {
+		strategy.ProgressivePerGroup = &RolloutProgressivePerGroup{}
+	}
+
+	// upgrade mandatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(strategy.ProgressivePerGroup.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
+
+	rolloutresult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(rolloutresult.ClustersToRollout) > 0 {
+		return &strategy, rolloutresult, nil
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.ProgressivePerGroup.Timeout.Timeout)
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// upgrade the rest decision groups
+	restClusterGroupKeys, restClusterGroups := r.pdTracker.ExistingClusterGroupsBesides(clusterGroupKeys)
+
+	rolloutresult = progressivePerGroup(restClusterGroupKeys, restClusterGroups, failureTimeout, statusFunc)
+	return &strategy, rolloutresult, nil
+}
+
+func progressivePerCluster(clusters []string, clusterToGroupKey map[string]clusterv1beta1.GroupKey, length int, timeout time.Duration, statusFunc ClusterRolloutStatusFunc) RolloutResult {
+	rolloutclusters := map[string]ClusterRolloutStatus{}
+	timeoutclusters := map[string]ClusterRolloutStatus{}
+
+	if length == 0 {
+		return RolloutResult{
+			ClustersToRollout: rolloutclusters,
+			ClustersTimeOut:   timeoutclusters,
+		}
+	}
+
+	// sort the clusters
+	sort.Strings(clusters)
```

Comment:
So the logic is to always progress in alphabetical order?

Reply:
yes, to ensure each time returns the same clusters that need to rollout.